### ISSUE TITLE
useSession on client return cached session instead of fetching

### ISSUE THREAD

**Issue by daveycodez** [better-auth/better-auth#986]
### Is this suited for github?

- [x] Yes, this is suited for github

### Describe the solution you'd like

If I'm using the useSession hook, whenever I refresh the page it sends an API request to get-session. We should already have the session locally, though. A comparison would be Supabase, their useSession hook does not send any requests unless the session is expired. Having to fetch the session when it already exists locally will make sites feel slower on initial load, and incur unnecessary bandwidth on the API. useSession needs to return the cached session on mount

### Describe alternatives you've considered

useSession hook return cached session immediately on client, only fetch get-session if it is expired.

**Comment by Bekacru** [comment]
The session cache is stored as an HTTP-only cookie, and the server needs to validate that the token is properly signed before sending back the session data. If there is no API request, it would mean storing the session in a way that js can read it, which isn’t ideal and is not recommended. Supabase do this mostly cause they don't have any other option since 3rd party cookies nowdays get blocked by browsers.

**Comment by daveycodez** [comment]
Makes sense. I'm also seeing that it's unsafe to store JWT in localStorage so I need to update some things there. Is it possible to have get-session return the jwt as well? I also saw examples of storing Bearer in localStorage, isn't that vulnerable to XSS attacks?

**Comment by Bekacru** [comment]
Yeah, it’s vulnerable to XSS, but it’s all about trade-offs. If an attacker manages to pull off an XSS attack, they can already send requests directly to your server, so stealing tokens might not even be even that important. They can do a lot of things. So, locking down your site against XSS should be the first priorit and it’s pretty straightforward these days.  And, JWTs can easily get too large for cookies, so local storage makes more sense. It’s not perfect, but as long as you’re aware of the risks and have proper security measures in place, it’s a workable option.

**Comment by daveycodez** [comment]
In this case wouldn't it also be fine to store the sessionData cached on localStorage so it can be used immediately on page load, so other requests (e.g. getMyProfile /api/profile/me) aren't dependent on waiting for that first getSession API request to resolve. Something like useOptimisticSession that returns it from localStorage and gets revalidated instead of starting as null.

Right now it goes like ->

**(Slow Network example on edit-profile page)**
1. Wait for Page Load (300ms)
2. Wait for getSession API Request (network latency - 300ms)
3. Get Current Profile from API (custom profiles sql table) (300ms)
Total time to render - 900ms

**Whereas if we had it cached it would be...**
1. Wait for Page Load (300ms)
2. Load cached sessionData (disk read latency) (~0ms)
3. Fire getSession (revalidate) and getCurrentProfile simultaneously (~300ms)
Total time to render - ~600ms


### Another example is headers with Avatars (image field).
1. Use cached sessionData (~0ms)
2. Render avatar (~0ms after page load) & revalidate getSession

**vs currently**
1. Wait for getSession(300ms)
2. Render Avatar (300ms time to render after page load)


I know I could make my own useOptimisticSession hook and cache it in localStorage but wouldn't it be better for all sites using better-auth to have faster perceived load times, or maybe even a config option for optimisticSessions (boolean)

**Comment by daveycodez** [comment]
The only concern I think with using localStorage is storing PII like email? But storing the JWT would also do that in its current state?

use-optimistic-session.ts
```ts
import { useEffect, useState } from "react"
import { useSession } from "@/lib/auth-client"

export function useOptimisticSession() {
    const { data: sessionData, isPending: sessionPending } = useSession()
    const [session, setSession] = useState(sessionData)
    const [isPending, setIsPending] = useState(true)

    // Load session from localStorage
    useEffect(() => {
        if (session) return

        const cachedSession = localStorage.getItem("session")

        if (cachedSession) {
            const sessionData = JSON.parse(cachedSession)
            if (new Date(sessionData.session.expiresAt) < new Date()) {
                localStorage.removeItem("session")
                return
            }

            setSession(sessionData)
            setIsPending(false)
        }
    }, [session])

    // Update session on revalidation
    useEffect(() => {
        if (isPending) setIsPending(sessionPending)
        if (sessionPending) return

        setSession(sessionData)

        if (sessionData) {
            localStorage.setItem("session", JSON.stringify(sessionData))
        } else {
            localStorage.removeItem("session")
        }
    }, [isPending, sessionData, sessionPending])

    return { session, isPending }
}
```

**Comment by Bekacru** [comment]
I think this should currently be handled by users rather than from the lib. Specifically, revalidating the cache becomes a problem. for example, if the user updates their info, the cookie cache will get overridden by the server. To do the same on the client, we would then need to track which endpoint is being calle and invalidate it which introduces a lot of complexity.

**Comment by daveycodez** [comment]
Since I'm just caching sessionData and refreshing it wouldn't it be the same outcome though? It still calls getSession on page load, which gets the latest session data and overrides the localStorage cache with the new data.

Another thing I think we need is a way to manually refresh the session coming from useSession hook. Like a refresh or reload function. In SWR they have revalidations trigger on page focus. Currently the useSession hook won't show you an updated email address or any other updated field to the user unless you do a hard refresh.

I definitely think default built in functionality for caching sessionData, revalidating on page focus, and proper retry on network failure is a necessity for any modern web application

**Comment by Bekacru** [comment]
If you need to control how session is cached and how and when refetch is triggered including onFocus and other advanced functionalities, you can use `getSession` with swr or react query instead. Said that, the session is cached in memory as it's with swr. Meaning calling `useSession` in two components doesn't trigger 2 different fetch requests. But local storage caching is a different story. We should probably mention using those libraries for advanced use cases for fetching session on the client.

**Comment by daveycodez** [comment]
Yea I know I can write the hooks, but I was kind of hoping to see this built in since everyone would benefit from the useSession hook having these features, both devs and all users of any site that has better-auth. Optimized out of the box without needing custom hooks on top of it. If there's no chance of this getting added to the library itself then I may just make a better-auth-swr package that can be used alongside it, but I feel like anyone using better-auth on a React based project definitely needs these features

**Comment by Bekacru** [comment]
Oh yeah, the library idea would be great. And if we had a dedicated React package, this would definitely make sense. The only thing we do for each framework (Vue, React, Svelte...) is provide a very lightweight adapter. For example, for the client, we just call the nanostore atoms with the respective `useStore` hooks and nothing else. In the future, we may have a dedicated adapter, but for now, since we're keeping `better-auth` as a single package, we try to avoid framework-specific code as much as possible.

**Comment by Nicolab** [comment]
Better-auth agnostic, it's great. I think also this should currently be handled by users (or another lib) rather than from better-auth lib.

I use Vue.js and React, the two have a different logic. If it can be useful to someone, here's how I do it with Vue.js:

```ts
watch(() => currentSession.value, (newVal) => {
    if (newVal.session.isPending || newVal.session.isRefetching) {
      appStore.user.value = null
      return
    }

    appStore.user.value = newVal?.user
}, { immediate: true })
```

By the way, from there, it's possible to synchronize the localStorage ;)

**Comment by daveycodez** [comment]
I don't think Vue uses the useSession React hook, the feature suggestion was for the React hook itself to reduce amount of "auth code" in the application. Plug and play with necessary features that every React app using better-auth would benefit from without having to re-write it manually and document it for every application. Yes we could all write our own logic to do these things. But we could also roll our own auth, right.

A feature like this is better to have built in than to Google through GitHub issues to figure out how to self implement, or just have a separate "official" better-auth-react library that might even be community managed. I think a lot of devs might not think of this issue even, so the users of those applications suffer the extra latency

**Comment by Nicolab** [comment]
UseSession() is a Nanostore, not a React hook. So Vue.js use also useSession() provided by Better-auth. 
Yes like you say, a different package for React would be relevant. That also open the door for Vue.js, Solid.js, Angular, htmx, Alpine, Astro, etc. So a list of external adapters is more clean, instead of bloat the main package with several useless dependencies.

**Comment by daveycodez** [comment]
In this case isn't localStorage agnostic, and can be used within useSession as an option to benefit every framework. Just a simple option in createAuthClient options enableLocalStorageCache Boolean. This doesn't add any dependencies and should only be a few lines of code inside of useSession and/or getSession and will make all sites in every framework using better-auth snappier on initial render

**Comment by reslear** [comment]
yeah that would be cool for Offline First (pwa/capacitor) applications
also i found wrap solution https://github.com/onejs/one/tree/af7e4049000f60e41a7b231d56f16d0393f72abc/packages/better-auth
but only for React but I think it could easily be adapted for vue.

**Comment by daveycodez** [comment]
@reslear 

https://github.com/daveyplate/better-auth-tanstack

I use my Tanstack library with a persistClient for offline auth now, in Capacitor

**Comment by reslear** [comment]
@daveycodez react 🤢

**Comment by joseph-9900** [comment]
@daveycodez Have you found any solution? Seeing the same result, on refresh the database is getting hit everytime! I have turned on cookieCache still

```
session: {
        expiresIn: 60 * 60 * 24 * 7, // 7 days
        updateAge: 60 * 60 * 24, // 1 day (every 1 day the session expiration is updated)
        cookieCache: {
            enabled: true,
            maxAge: 5 * 60 // Cache duration in seconds
        }
    },
```

**Comment by Nikola-Milovic** [comment]
Any news? I want to have offline capabilities and it would be beneficial to have session persist even if network is not available. Can this be solved via the plugin system?

**Comment by daveycodez** [comment]
> Any news? I want to have offline capabilities and it would be beneficial to have session persist even if network is not available. Can this be solved via the plugin system?

https://github.com/daveyplate/better-auth-tanstack

Use this with a TS Query Persist Client https://tanstack.com/query/latest/docs/framework/react/plugins/persistQueryClient
